### ISSUE TITLE
Allow for a SYSGPORT configuration statement

### ISSUE THREAD

**Issue by arfineman** [SDL-Hercules-390/hyperion#505]
It would be nice to have a SYSGPORT statement to distinguish a SYSG port from a 3270 port. If not specified, it would default to the current CNSLPORT.

**Comment by Fish-Git** [comment]
Why? What's the problem? Why would you want/need to use a separate port?

**Comment by arfineman** [comment]
Currently if you have a SYSG defined, the first 3270 connection
automatically gets assigned to SYSG. Then any subsequent connections are
assigned to 3270 devices. But this may not be desirable. On real hardware,
SYSG is started on demand and one could have lots of 3270 devices without
SYSG.
Best regards,

**Comment by Fish-Git** [comment]
> Currently, if you have a SYSG defined, the first 3270 connection automatically gets assigned to SYSG. Then any subsequent connections are assigned to 3270 devices.

Only if your configuration file is coded to behave that way.

> But this may not be desirable. On real hardware, SYSG is started on demand and one could have lots of 3270 devices without SYSG.

Understood. Which is why Hercules supports the `groupname`, `ipaddr` and `mask` options on its 3270 device statements, including SYSG devices:

* https://sdl-hercules-390.github.io/html/hercconf.html#3270
* https://sdl-hercules-390.github.io/html/hercconf.html#SYSG

**Comment by Fish-Git** [comment]
_**P.S.** &nbsp; It is preferred that you  **not** respond/reply to GitHub Issues via email **`(*)`**, and instead respond/reply _directly_ via the GitHub Issues web page itself:_

* **[ht<span>tps</span>://github.com/SDL-Hercules-390/hyperion/issues](https://github.com/SDL-Hercules-390/hyperion/issues)**

_When you reply _directly_ via their web page, I can make minor edits to your reply so it is more readable (prettier) by editing the fonts being used, formatting the log messages, etc. When you reply via email however, I cannot edit your reply (GitHub does not allow it), so oftentimes it is much harder (more difficult) to read._

_It is up to you whether or not you want to take the time to reply via their web page or continue to reply via email, but it is generally preferable that you reply directly via their web page instead._

_Thanks_

_**(\*)** GitHub does not support formatting of email replies, making it impossible for me to fix the formatting of a person's reply for readability. Thank you for understanding._

**Comment by arfineman** [comment]
Thank you, Fish.
The problem maybe with my 3270 emulator. I can't specify a groupname or a device address on the hostname.
I maybe wrong, but looks like the doc implies that not all emulators support that. 
Currently I just comment out the SYSG and only put it back in when needed. 
Again, I thought this is a nice to have, but if you feel its not needed, you can close it out.
Best regards,

**Comment by Fish-Git** [comment]
> The problem may be with my 3270 emulator. I can't specify a groupname or a device address on the hostname.

What emulator are you using?

> Again, I thought this is a nice to have, but if you feel its not needed, you can close it out.

It certainly has merit, I'll give you that. And it should certainly be easy enough to implement too IMO, so I think we should probably keep this request open. Maybe someone out there will be able to implement it for you before I do. (*)

 Any takers?
___
(*) _I'd do it myself but I'm kind of busy with the "COD" thing right now, so if I do it it'll have to wait until later. Sorry. But maybe somebody else can do it before I can? Hello? Anyone?_

**Comment by Fish-Git** [comment]
Aaron (@arfineman),

Closed by commit 6dee8ff37b3b92a10fc33d873c5f6eea2cce5f8c.

Please git it a try and tell me what you think.

Thanks.

**Comment by arfineman** [comment]
Hi Fish,
Thank you so much for putting this in. I'm sure besides me a lot of other people will make good use of this feature. This allows users to create a separate x3270 session for the SYSG port. I'm a Windows user and never had success building Hercules from source. So I'll have to wait for the next release, which I hope will come out soon.
Best regards,

**Comment by Fish-Git** [comment]
> I'm a Windows user

As am I.

> and never had success building Hercules from source.

What's the problem? Building Hercules from source is actually pretty easy IMO. Of course I'm a developer too and so do it all the time which is why I find it to be easy. But even for the average non-developer Windows user it's not really _that_ difficult.

In fact, Bill Lewis (i.e. @wrljet, a fellow Hercules developer) has created a very nice, easy to use Windows Power Shell helper script called "**[Hercules Helper](https://github.com/wrljet/hercules-helper-windows)**" that totally automates the entire process, making doing so dirt simple. You just enter a simple command, press enter, and in a few minutes, _**Voilà!**_ You've got a working version of Hercules on your system. &nbsp;`:)`

Of course you have to clone his repository first, which requires having `git` already installed on your system, but that's not really a big problem. Installing `git` is just a simple download and a few mouse clicks, and cloning Bill's "[Hercules Helper](https://github.com/wrljet/hercules-helper-windows)" repository is just a simple one-line command once `git` is installed. _&lt;shrug&gt;_

Try it! &nbsp; I think you'll be glad you did.

(Bill? Does your script support building the "develop" branch?)

**Comment by wrljet** [comment]
"develop" branch is the default.

You can download the [Hercules-Helper-Windows](https://github.com/wrljet/hercules-helper-windows/archive/refs/heads/master.zip) repo as a zip file from GitHub, without git.

I'm here, and all ears, to help anyone who wants it.

Bill

**Comment by Fish-Git** [comment]
> "develop" branch is the default.

That's good to hear!

> You can download the [Hercules-Helper-Windows](https://github.com/wrljet/hercules-helper-windows/archive/refs/heads/master.zip) repo as a zip file from GitHub, without git.

_(Doh!)_ &nbsp;Of course. &nbsp;How silly of me. &nbsp;`%-P`

**Comment by wrljet** [comment]
I have also just created a macOS Homebrew installation formula.
Makes installing the latest SDL-Hercules-390 develop absolutely trivial on macOS.
(not "released" yet)

**Comment by arfineman** [comment]
Thank you, Bill. I downloaded the zip file. But neither my W7 or W10 can open the 'README.md' to start the process. Am I doing something wrong, or do I need to download something to read that file. I do have Office 2016 installed.
Best regards,

**Comment by arfineman** [comment]
Please ignore the previous message. Looks like if I associate .md file extension to Wordpad, it opens fine. My bad.

**Comment by wrljet** [comment]
Aaron,

`README.md` is a "Markdown" file.  That is the default for such things on GitHub.

If you scroll down on the [GitHub page](https://github.com/wrljet/hercules-helper-windows) past the file list, it displays the `README.md` with its formatting.

I don't know how to view those with their formatting on Windows.

Bill

**Comment by Fish-Git** [comment]
> But neither my W7 or W10 can open the 'README.md' to start the process.

What do you mean by "cannot open"? Did you "unblock" the downloaded .zip file before you unzipped it, [as Bill's web site explains in Step 2](https://github.com/wrljet/hercules-helper-windows)?

> _Sometimes when files are copied to Windows, it will decide they have come from an untrusted remote system, and a directory bit will be set to remember this fact. This will cause Windows to not allow them to be executed._
> 
> _Right-click on the hercules-helper.zip file and select Properties. At the bottom ofthe Properties dialog, you may see a warning about the file coming from a remote system. Click the button to Unblock the file. Close this dialog._

This is standard Windows behavior:

* https://thirtysix.zendesk.com/hc/en-us/articles/202921675-How-to-Unblock-a-File-Downloaded-from-an-Email-or-the-Internet
* https://support.gfi.com/hc/en-us/articles/360012969033-Microsoft-Windows-Blocks-Downloaded-GFI-Software-File
* https://winaero.com/how-to-unblock-files-downloaded-from-the-internet-in-windows-11/

I believe ".md" files are treated as browser files, just like ".htm" or ".html" files are, and thus either your browser or Windows itself may be purposely preventing you from opening them since it considers them "untrustworthy" since they came from a foreign system. You need to manually unblock the file before attempting to open it.

As far as _displaying_ the README.md files in an easy-to-read format in your browser, you might consider installing a "markdown viewer" type extension, such as the following one that I myself use for Chrome:

* https://chrome.google.com/webstore/detail/markdown-viewer/ckkdlimhmcjmikdlpkmbgfkaikojcbjk?hl=en

If you so search the web there are even "online" web pages where you can upload or paste your markdown file/code into a text box and it will display the rendered markdown code for you.

_(Markdown files are just simple, specially formatted text files, similar to HTML files that can also be opened in Notepad, but which when opened in a browser are parsed and displayed in a much more user friendly manner.)_

**Comment by Fish-Git** [comment]
But all of that _(what I've written in my previous comment above)_ is only if you want to view on your own system the information that already exists on Bill's web site:

* https://github.com/wrljet/hercules-helper-windows

Visit the above web page and scroll down maybe one page, and you will see _exactly_ the same information that is in the README.md file. The GitHub web site automatically displays .md files already properly formatted for you. If you want to display the same thing (i.e. .md files) _locally_ in your own browser window, then you need to install a browser extension like the one I suggested.

**Comment by Fish-Git** [comment]
> I don't know how to view those with their formatting on Windows.

You need to use a browser extension, such as [Markdown Viewer](https://chrome.google.com/webstore/detail/markdown-viewer/ckkdlimhmcjmikdlpkmbgfkaikojcbjk?hl=en).

**Comment by Fish-Git** [comment]
> Looks like if I associate .md file extension to Wordpad, it opens fine.

You _shouldn't_ need to do that! You _should_ (IMHO) leave the file associated with your browser, just like it was. Opening and properly displaying .md files is not that hard to do once you know how.

_(And once you UNBLOCK them beforehand, of course!)_

**Comment by wrljet** [comment]
> > I don't know how to view those with their formatting on Windows.
> 
> You need to use a browser extension, such as [Markdown Viewer](https://chrome.google.com/webstore/detail/markdown-viewer/ckkdlimhmcjmikdlpkmbgfkaikojcbjk?hl=en).

I don't *do* Chrome.

**Comment by Fish-Git** [comment]
> I don't _do_ Chrome.

What _do_ you do?

**Comment by wrljet** [comment]
Thanks, Fish!  I put in the one for MSFT Edge, and it just worked.

**Comment by arfineman** [comment]
Hi Bill,
Does hyperion-buildall.ps1 script download the Hercules source, or do I need to do that manually ?
If so, is there a wget command to get the latest source ?
Best regards,

**Comment by wrljet** [comment]
It will install Visual Studio (you can select which major version you prefer) with the required workloads, clone the SDL-Hercules-390 GitHub repo, and do the whole thing.

(That said, I don't often retest with a completely pure virgin Windows, so things can break over time.)

Be sure to add the `-GitBranch develop` option to the command line so you get the latest with the new feature.

**Comment by arfineman** [comment]
There has to be powershell for dummies manual somewhere.

PS C:\users\asus\desktop\hrc\hrc> .\hyperion-buildall
File C:\users\asus\desktop\hrc\hrc\hyperion-buildall.ps1 cannot be loaded. The file C:\users\asus\desktop\hrc\hrc\hyper
ion-buildall.ps1 is not digitally signed. The script will not execute on the system. Please see "get-help about_signing
" for more details..
At line:1 char:20
+ .\hyperion-buildall <<<<
    + CategoryInfo          : NotSpecified: (:) [], PSSecurityException
    + FullyQualifiedErrorId : RuntimeException

PS C:\users\asus\desktop\hrc\hrc>

**Comment by wrljet** [comment]
I hate PowerShell as much as the next guy, but it's what Windows gives us.

You need to enable it to run foreign untrusted scripts.

Check out my [Step 1](https://github.com/wrljet/hercules-helper-windows#step-1) where it talks about `Set-ExecutionPolicy RemoteSigned`.

**Comment by wrljet** [comment]
OK
This is where I normally start cussing.  

Let's start with, what Windows are you on, and what version of PowerShell is it?

With that info I'll fire up a fresh copy in a VM.

Bill

**Comment by wrljet** [comment]
One more question: did you close and reopen the PowerShell window after doing `Set-ExecutionPolicy RemoteSigned`?

**Comment by arfineman** [comment]
OK. Making progress. I exited PS and came back in, 
Now it is complaining about a -BuildDir paramater.
Where do I specify -Gitbranch develpp ? 
Can you post the exact command I need to enter.
Best reagrds,

PS C:\users\hp\desktop\hrc\hrc> .\hyperion-buildall.ps1
Transcript started, output file is .\hercules-helper-20221018_18-56-49.log

Options:
C:\users\hp\desktop\hrc\hrc\hyperion-buildall.ps1 : Error: -BuildDir parameter is missing
At line:1 char:1
+ .\hyperion-buildall.ps1
+ ~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
    + FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException,hyperion-buildall.ps1

Transcript stopped, output file is C:\users\hp\desktop\hrc\hrc\hercules-helper-20221018_18-56-49.log
PS C:\users\hp\desktop\hrc\hrc> .\hyperion-buildall.ps1 -gitbranch develop
Transcript started, output file is .\hercules-helper-20221018_18-58-58.log

Options:
C:\users\hp\desktop\hrc\hrc\hyperion-buildall.ps1 : Error: -BuildDir parameter is missing
At line:1 char:1
+ .\hyperion-buildall.ps1 -gitbranch develop
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
    + FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException,hyperion-buildall.ps1

Transcript stopped, output file is C:\users\hp\desktop\hrc\hrc\hercules-helper-20221018_18-58-58.log
PS C:\users\hp\desktop\hrc\hrc> .\hyperion-buildall.ps1 -gitbranch develop

**Comment by Fish-Git** [comment]
> Now it is complaining about a `-BuildDir` parameter. Where do I specify `-Gitbranch develop`? Can you post the exact command I need to enter?

It's all documented in Bill's README!

Refer to the instructions for Step 4:

* https://github.com/wrljet/hercules-helper-windows#step-4

Decide _where_ (what directory) you want Hercules to be built, and decide if you prefer to use Visual Studio 2017, 2019, or 2022, and then specify those options (as well as the -GitBranch option too) in your command, e.g:

        .\hyperion-buildall.ps1 -VS2017 -BuildDir c:\hercules -GitBranch develop

**Comment by arfineman** [comment]
Thank you. I misunderstood. I thought that script does all those steps for you.
Best regards,

**Comment by arfineman** [comment]
Success. This script is unbelievable. Build is complete. Now I can pickup the latest and greatest without having to wait for a new official release. Not I can test the SYSGPORT command.
Thanks so much to Bill and Fish.

**Comment by wrljet** [comment]
Glad to hear it!

Aaron, if you don't mind, would you find the log file it created, the one with the newest timestamp, zip it and send to me.
I'd like to review it.  The filename will be along the lines of `hercules-helper-20220920_18-26-13.log`.

**Comment by arfineman** [comment]
Will do. This was done on my home PC as will take care of it this evening when I get home. Thanks again for putting this together for dummies like me. I had tried it several times before on my own without success. 
Best regards,

**Comment by arfineman** [comment]
Hi Fish,
I tested the SYSGPORT as a configuration statement and also as a command and it works perfectly. I see that you set the default port number to 3278 that works great for me. I added a SYSGPORT 3278 to my config file regardless. But I just thought this may be problematic for an existing user of SYSG that is used to the old port 3270 and this change may not be considered backward compatible. This is strictly your call if you want the default to be 3270 or 3278.   
Best regards,

**Comment by Fish-Git** [comment]
> I see that you set the default port number to 3278. [...] I should point out however that this may be problematic for an existing user of SYSG that is used to the old port 3270, as this change may not be considered backward compatible. This is strictly your call if you want the default to be 3270 or 3278.

Unfortunately the `SYSGPORT` port number _must_ be different from the `CNSLPORT` port number. This is because you can't have two listening sockets both listening for connections on the same port. Otherwise when someone tries connecting to that port, it would be completely unpredictable which listening socket would end up accepting the connection (thereby making it impossible to determine whether the user wanted to connect a SYSG console or a regular 3270 terminal). The new code _should_ be enforcing this restriction. Otherwise it just wouldn't work right. You can verify this for me if you want by trying to define both of them to the same port.  Depending on which one you define first, the other one _should_ be rejected with an error.

Your point _is_ valid however: introducing new unexpected behavior is rarely a Good Thing.

I suppose I _could_ tweak the new code a tiny bit to support specifying e.g. `SYSGPORT NO` (and make that the default) to prevent the new SYSG port listening socket from being created. Doing that (and making a few other minor adjustments) _should_ I believe allow the old behavior to work the same as it did before.

Let me think about it...

**Comment by arfineman** [comment]
Log attached.

**Comment by Fish-Git** [comment]
> Log attached.

Um, no it isn't. &nbsp;`:(`

This is yet another reason why [_I prefer that people respond directly to GitHub Issues and **not** via email._](https://github.com/SDL-Hercules-390/hyperion/issues/505#issuecomment-1261455127)

**Comment by arfineman** [comment]
Sorry, I replied this time, expecting it directly to go to Bill. Here is the attachment.
[hercules-helper-20221018_20-09-00.zip](https://github.com/SDL-Hercules-390/hyperion/files/9826048/hercules-helper-20221018_20-09-00.zip)

**Comment by Fish-Git** [comment]
> ... expecting it directly to go to Bill.

That would only occur if you manually changed the email "To:" field to his email address. If you don't do that, the email goes to GitHub.

**Comment by arfineman** [comment]
Fish,
My opinion on this is that your update fixed an existing problem. On a real hardware, SYSG is only used either during IPL or an emergency when TCP/IP is lost. It is known as the Integrated 3270 Console and is a standard feature on HMC and is only started on demand. The way this was implemented on Hercules, you were forced to open a SYSG session if you wanted to use any of the 3270 ports. This was inconsistent with the way it works on HMC, where OSA-ICC (Local 3270) devices connect independently. I know you initially marked it as an enhancement, but I considered it a bug. I was just happy that you agreed to consider it.
Best regards,

**Comment by Fish-Git** [comment]
> Your point _is_ valid however: introducing new unexpected behavior is rarely a Good Thing.
> 
> I suppose I _could_ tweak the new code a tiny bit to support specifying e.g. `SYSGPORT NO` (and make that the default) to prevent the new SYSG port listening socket from being created. Doing that (and making a few other minor adjustments) _should_ I believe allow the old behavior to work the same as it did before.
> 
> Let me think about it...

_**Done!**_

Aaron? (@arfineman) Please rebuild and re-test. The original Hercules behavior (before any of my changes) should be restored now when a SYSGPORT statement is not specified (or is specified as "NO"). The new behavior is only triggered whenever a SYSGPORT statement is specified with a valid port number.

This _should_ prevent "unexpected new behavior" for our existing users while still allowing them to take advantage of the new behavior should they wish to.

Thanks.

I am closing this issue as completed again. This time hopefully for good! &nbsp;`;-)`

**Comment by wrljet** [comment]
Aaron,

Hercules-Helper will not re-clone or update your local SDL-Hercules-390 git repo.
This is for safety, until I add an override option for that.

So to update and rebuild:

```
pushd C:\hercules\hyperion
git pull
popd
```

The Visual Studio installation should have left you with a command line `git`, so the above will hopefully just work.

Then rerun your original hyperion-buildall command.

Bill

**Comment by arfineman** [comment]
Fish,
I tested the changes and it's perfect. It just can't get any better than this. Without SYSGPORT it behaves exactly as before, as you intended. With SYSGPORT specified, it acts just like a z15. Thank again again for fixing this. And thank you Bill for providing the wonderful Hercules_Helper tool.
Best regards, Aaron

**Comment by wrljet** [comment]
Fish,

The latest commit https://github.com/SDL-Hercules-390/hyperion/commit/66ff70ab4830660fef2a881ba4918fdb212972cf , related to this SYSGPORT, is busted (on Linux).

Hercules won't connect to the 3270 console with `c3270 127.0.0.1:3270`.  It just hangs, waiting.
This was reported to me on the Moshix Discord channel and I have reproduced it here on Debian 10 and 11, and narrowed it down to the latest commit.

Bill

**Comment by Fish-Git** [comment]
> The latest commit [66ff70a](https://github.com/SDL-Hercules-390/hyperion/commit/66ff70ab4830660fef2a881ba4918fdb212972cf) , related to this SYSGPORT, is busted (on Linux).
> 
> Hercules won't connect to the 3270 console with `c3270 127.0.0.1:3270`. It just hangs, waiting. This was reported to me on the Moshix Discord channel and I have reproduced it here on Debian 10 and 11, and narrowed it down to the latest commit.

That's odd. The logic is platform independent and should work identically on *Nix as it does on Windows. I promise you I _did_ test it on Windows, and it worked just fine! Aaron, who also uses Windows, also reported that it worked just fine for him too, so why the frick it's failing on *nix I have no freaking idea.

(sigh) Let me look into it an get back to you... &nbsp;`:(`

**Comment by Fish-Git** [comment]
_(oops!)_ How the frick did _that_ happen? Sorry! Should be fixed now.

**Comment by wrljet** [comment]
Thanks.

Is that ( https://github.com/SDL-Hercules-390/hyperion/commit/44dc0f8c7cf2efe4081bade9fe190880215e06ac) really the title/comment you want in that commit?

**Comment by Fish-Git** [comment]
> Is that ( [44dc0f8](https://github.com/SDL-Hercules-390/hyperion/commit/44dc0f8c7cf2efe4081bade9fe190880215e06ac)) really the title/comment you want in that commit?

Sure! Why not? It was just a fix of a silly typo. No need to get into details IMO.
